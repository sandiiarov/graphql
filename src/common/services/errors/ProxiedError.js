// @flow

type ProxiedErrorValues = {|
  +statusCode: string,
  +url: string,
|};

/**
 * This exception is thrown if origin server does NOT respond with HTTP status code 200.
 */
class ProxiedError extends Error {
  extensions: {|
    +_proxy: ProxiedErrorValues,
    +extensions: {|
      +proxy: ProxiedErrorValues,
    |},
  |};

  constructor(message: string, originStatusCode: string, originUrl: string) {
    super(message);
    this.extensions = {
      _proxy: {
        // note: this is deprecated and will be removed
        statusCode: originStatusCode,
        url: originUrl,
      },
      extensions: {
        proxy: {
          statusCode: String(originStatusCode),
          url: originUrl,
        },
      },
    };
  }
}

export { ProxiedError };
