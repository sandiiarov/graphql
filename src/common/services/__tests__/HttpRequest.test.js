// @flow

import { get, post } from '../HttpRequest';
import { ProxiedError } from '../errors/ProxiedError';

jest.mock('node-fetch');

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  process.env.NODE_ENV = 'production';
});

afterEach(() => {
  jest.resetAllMocks();
  process.env.NODE_ENV = 'test';
});

describe('GET request', () => {
  it('should throw exception in test environment', async () => {
    process.env.NODE_ENV = 'test';
    await expect(get('https://path/to/api')).rejects.toBeError(
      'HttpRequest should never be called in test environment. Have you forgotten to mock "https://path/to/api" with fake data response?',
    );
  });
});

describe('GET request in production', () => {
  it('resolves URL with token', async () => {
    expect(await get('https://path/to/api', 't/ok"en')).toMatchSnapshot();
  });

  it('resolves URL without token', async () => {
    expect(await get('https://path/to/api')).toMatchSnapshot();
  });

  it('resolves URL with query string and token', async () => {
    expect(
      await get('https://path/to/api?locale=en', 'randomToken'),
    ).toMatchSnapshot();
  });

  it('throws exception during invalid return status code', async () => {
    expect.assertions(3);

    try {
      await get('https://path/to/api?status=500');
    } catch (error) {
      expect(error).toBeInstanceOf(ProxiedError);
      expect(error.message).toBe('Status Text');
      expect(error.extensions).toMatchSnapshot();
    }
  });

  it('contains proper user agent header', async () => {
    const resp = await get('https://path/to/api');
    expect(resp.options.headers['X-Client']).toEqual('graphql');
  });
});

describe('POST request', () => {
  it('should throw exception in test environment', async () => {
    process.env.NODE_ENV = 'test';
    await expect(post('https://path/to/api', {})).rejects.toBeError(
      'HttpRequest should never be called in test environment. Have you forgotten to mock "https://path/to/api" with fake data response?',
    );
  });
});

describe('POST request in production', () => {
  it('throws exception during invalid return status code', async () => {
    expect.assertions(3);

    try {
      await get('https://path/to/api?status=500');
    } catch (error) {
      expect(error).toBeInstanceOf(ProxiedError);
      expect(error.message).toBe('Status Text');
      expect(error.extensions).toMatchSnapshot();
    }
  });

  it('adds default content type header', async () => {
    expect(await post('https://path/to/api', {})).toMatchSnapshot();
  });

  it('supports custom content type header', async () => {
    expect(
      await post(
        'https://path/to/api',
        {},
        { 'Content-Type': 'application/soap+xml' },
      ),
    ).toMatchSnapshot();
  });

  it('supports custom headers', async () => {
    expect(
      await post(
        'https://path/to/api',
        {},
        {
          'Content-Type': 'application/soap+xml',
          'X-Custom-A': 'Value A',
          'X-Custom-B': 'Value B',
        },
      ),
    ).toMatchSnapshot();
  });

  it('stringify payload', async () => {
    expect(
      await post('https://path/to/api', {
        email: 'john@example.com',
        password: 123456,
      }),
    ).toMatchSnapshot();
  });

  it('contains proper user agent header', async () => {
    const resp = await get('https://path/to/api');
    expect(resp.options.headers['X-Client']).toEqual('graphql');
  });
});
