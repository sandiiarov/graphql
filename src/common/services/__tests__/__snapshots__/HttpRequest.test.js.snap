// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET request in production resolves URL with query string and token 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "headers": Object {
      "Content-Type": "application/json",
      "X-Client": "graphql",
    },
    "method": "GET",
  },
  "url": "https://path/to/api?locale=en&token=randomToken",
}
`;

exports[`GET request in production resolves URL with token 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "headers": Object {
      "Content-Type": "application/json",
      "X-Client": "graphql",
    },
    "method": "GET",
  },
  "url": "https://path/to/api?token=t%2Fok%22en",
}
`;

exports[`GET request in production resolves URL without token 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "headers": Object {
      "Content-Type": "application/json",
      "X-Client": "graphql",
    },
    "method": "GET",
  },
  "url": "https://path/to/api",
}
`;

exports[`GET request in production throws exception during invalid return status code 1`] = `
Object {
  "_proxy": Object {
    "statusCode": "500",
    "url": "https://path/to/api?status=500",
  },
  "extensions": Object {
    "proxy": Object {
      "statusCode": "500",
      "url": "https://path/to/api?status=500",
    },
  },
}
`;

exports[`POST request in production adds default content type header 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "body": "{}",
    "headers": Object {
      "Content-Type": "application/json",
      "X-Client": "graphql",
    },
    "method": "POST",
  },
  "url": "https://path/to/api",
}
`;

exports[`POST request in production stringify payload 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "body": "{\\"email\\":\\"john@example.com\\",\\"password\\":123456}",
    "headers": Object {
      "Content-Type": "application/json",
      "X-Client": "graphql",
    },
    "method": "POST",
  },
  "url": "https://path/to/api",
}
`;

exports[`POST request in production supports custom content type header 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "body": "{}",
    "headers": Object {
      "Content-Type": "application/soap+xml",
      "X-Client": "graphql",
    },
    "method": "POST",
  },
  "url": "https://path/to/api",
}
`;

exports[`POST request in production supports custom headers 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "body": "{}",
    "headers": Object {
      "Content-Type": "application/soap+xml",
      "X-Client": "graphql",
      "X-Custom-A": "Value A",
      "X-Custom-B": "Value B",
    },
    "method": "POST",
  },
  "url": "https://path/to/api",
}
`;

exports[`POST request in production throws exception during invalid return status code 1`] = `
Object {
  "_proxy": Object {
    "statusCode": "500",
    "url": "https://path/to/api?status=500",
  },
  "extensions": Object {
    "proxy": Object {
      "statusCode": "500",
      "url": "https://path/to/api?status=500",
    },
  },
}
`;
